cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -1565,5 +1589,43 @@ pub fn define(
     e.enc_both(ffcmp.bind(F32), rec_fcmp.opcodes(vec![0x0f, 0x2e]));
     e.enc_both(ffcmp.bind(F64), rec_fcmp.opcodes(vec![0x66, 0x0f, 0x2e]));
 
+    // SIMD splat: before x86 can use vector data, it must be moved to XMM registers; see legalize.rs for
+    // how bitcast is used to move the data there
+
+    let mut shuffle_mapping: HashMap<u64, Vec<u8>> = HashMap::new();
+    shuffle_mapping.insert(8, vec![0x66, 0x0f, 0x38, 00]); // PSHUFB from SSSE3, TODO needs another instruction to get the mask into another register (plus bitcast)
+    // TODO legalize 16x8 to two PINSRW (from SSE), then PSHUFD as if it were a 32x4?
+    shuffle_mapping.insert(32, vec![0x66, 0x0f, 0x70]); // PSHUFD from SSE2
+    // TODO legalize 64x2 as two PINSRQ instructions (from SSE 4.1)?
+
+    for ty in LaneTypeIterator::new() {
+        if let Some(opcode) = shuffle_mapping.get(&ty.lane_bits()) {
+            let number_of_lanes = 128 / ty.lane_bits();
+            e.enc32(x86_pshuf.bind_vector(ty, number_of_lanes), rec_r_ib_unsigned.nonrex().opcodes(opcode.clone()));
+            e.enc64(x86_pshuf.bind_vector(ty, number_of_lanes), rec_r_ib_unsigned.nonrex().opcodes(opcode.clone()));
+            // TODO can't use enc_both here because it forces a .rex() encoding
+        }
+    }
+
+    // SIMD insertlane
+
+    // PINSRB, 8-bit insertion into a lane requiring SSE4.1
+    e.enc_32_64(insertlane.bind_vector(B8, 16), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0x3a, 0x20]));
+    e.enc_32_64(insertlane.bind_vector(I8, 16), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0x3a, 0x20]));
+
+    // PINSRW, 16-bit insertion into a lane requiring SSE2
+    e.enc_32_64(insertlane.bind_vector(B16, 8), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0xc4]));
+    e.enc_32_64(insertlane.bind_vector(I16, 8), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0xc4]));
+
+    // PINSRD, 32-bit insertion into a lane requiring SSE4.1
+    e.enc_32_64(insertlane.bind_vector(B32, 4), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
+    e.enc_32_64(insertlane.bind_vector(F32, 4), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
+    e.enc_32_64(insertlane.bind_vector(I32, 4), rec_r_ib_unsigned_r.nonrex().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
+
+    // PINSRQ, 64-bit insertion into a lane requiring SSE4.1 (only on x86_64)
+    e.enc64(insertlane.bind_vector(B64, 4), rec_r_ib_unsigned_r.rex().w().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
+    e.enc64(insertlane.bind_vector(F64, 2), rec_r_ib_unsigned_r.rex().w().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
+    e.enc64(insertlane.bind_vector(I64, 2), rec_r_ib_unsigned_r.rex().w().opcodes(vec![0x66, 0x0f, 0x3a, 0x22]));
```

Comment:
I had to start implementing some of the `insertlane` instructions in order to proceed with my current legalizations.